examples/tic_tac_toe/tic-tac-toe.py

```diff
@@ -0,0 +1,184 @@
+
+        requested_at = int(time.time() * 1000)
+        messages = trajectory.messages()
+
+        try:
+            client = model.openai_client()
+            chat_completion = await client.chat.completions.create(
+                model=model.name,
+                messages=messages,
+                max_completion_tokens=128,
+            )
+            last_completion = chat_completion
+        except openai.LengthFinishReasonError as e:
+            raise e
+        except Exception as e:
+            print("caught exception generating chat completion")
+            print(e)
+            global failing_trajectory
+            failing_trajectory = trajectory
+            raise e
+
+        try:
+            op_client.report(
+                requested_at=requested_at,
+                received_at=int(time.time() * 1000),
+                req_payload={
+                    "messages": messages,
+                    "metadata": {
+                        "notebook-id": "tic-tac-toe",
+                        "iteration": str(iteration),
+                        "validation": str(is_validation),
+                        "move_number": str(move_number),
+                    },
+                },
+                resp_payload=chat_completion,
+                status_code=200,
+            )
+        except Exception as e:
+            print(f"Error reporting to OpenPipe: {e}")
+
+        choice = chat_completion.choices[0]
+        content = choice.message.content
+        assert isinstance(content, str)
+        trajectory.messages_and_choices.append(choice)
+
+        try:
+            apply_agent_move(game, content)
+        except ValueError as e:
+            trajectory.reward = -1 + (math.log(move_number + 1) / math.log(100))
+            break
+
+        if check_winner(game["board"]) is not None:
+            break
+        move_number += 1
+
+        opponent_move = get_opponent_move(game)
+        game["board"][opponent_move[0]][opponent_move[1]] = game["opponent_symbol"]
+
+    winner = check_winner(game["board"])
+
+    if winner == game["agent_symbol"]:
+        trajectory.reward = 1
+    elif winner == game["opponent_symbol"]:
+        trajectory.reward = 0
+    elif winner == "draw":
+        trajectory.reward = 0.5
+
+    try:
+        op_client.update_log_metadata(
+            filters=[
+                {
+                    "field": "completionId",
+                    "equals": last_completion.id,
+                }
+            ],
+            metadata={
+                "reward": str(trajectory.reward),
+                "reward_assigned": "true",
+            },
+        )
+    except Exception as e:
+        print(f"Error updating log metadata: {e}")
+
+        print(trajectory.reward)
+
+    return trajectory
+
+
+DESTROY_AFTER_RUN = False
+
+
+async def main():
+    # run from the root of the repo
+    api = art.SkyPilotAPI(cluster_name="art6")
+    await api.initialize_cluster(art_version=".", env_path=".env")
```

Comment:
after looking at this I still feel it's too magical and I'd like us to add a required parameter named `gpu`, just like Modal has (if you want to run on GPU). I feel like as a user I'll have a much better mental model of what this line is doing if I explicitly have to put in that I'm provisioning a GPU here.

Reply:
GPU could be optional if you pass in a `resources=` block instead.